core/system/src/secure_gateway.c

```diff
@@ -0,0 +1,250 @@
+ * @param svc_id[in]    Full ID field of the SVCall opcode */
+bool secure_gateway_context_switch_in(uint32_t src_svc_sp, uint32_t src_svc_pc, uint8_t src_svc_id)
+{
+    int nargs;
+    uint8_t dst_id;
+    uint32_t dst_fn;
+    uint32_t src_sp, dst_sp;
+    uint32_t xpsr;
+    TSecureGateway * secure_gateway;
+
+    nargs = (int) UVISOR_SVC_FAST_NARGS_GET(src_svc_id);
+    if (nargs > CONTEXT_SWITCH_FUNCTION_MAX_NARGS) {
+        HALT_ERROR(NOT_ALLOWED, "Max 4 32 bit arguments allowed");
+    }
+
+    /* The stack pointer provided to us as an input comes from an SVC exception.
+     * We must check that it corresponds to a full frame that the source box can
+     * access. */
+    /* This function halts if it finds an error. */
+    src_sp = context_validate_exc_sf(src_svc_sp);
+
+    /* Check that the SVC came from within a secure gateway. */
+    /* This function halts if it finds an error. */
+    secure_gateway = secure_gateway_check((TSecureGateway *) src_svc_pc);
+
+    /* Destination box: Gather information from the secure gateway. */
+    dst_id = secure_gateway_id_from_cfg_ptr(secure_gateway->dst_box_cfg_ptr);
+    dst_fn = secure_gateway->dst_fn;
+
+    /* Currently we do not support secure gateways to oneself. Secure gateways
+     * to box 0 are also not allowed. */
+    if (!dst_id) {
+        HALT_ERROR(NOT_ALLOWED, "A secure gateway to box 0 is not allowed.");
+    }
+    if (g_active_box == dst_id) {
+        HALT_ERROR(NOT_ALLOWED, "The destination (%d) of a secure gateway must be different from its source (%d).",
+                                dst_id, g_active_box);
+    }
+
+    /* Forge a stack frame for the destination box and transfer nargs arguments
+     * from the source stack. */
+    xpsr = ((uint32_t *) src_sp)[7];
+    dst_sp = context_forge_exc_sf(src_sp, dst_id, dst_fn, (uint32_t) secure_gateway_thunk, xpsr, nargs);
+
+    /* Perform the context switch-in to the destination box. */
+    /* This function halts if it finds an error. */
+    context_switch_in(CONTEXT_SWITCH_FUNCTION_GATEWAY, dst_id, src_sp, dst_sp);
+
+    /* Return whether the destination box requires privacy or not. */
+    /* TODO: Context privacy is currently unsupported. */
+    return false;
+}
+
+/** Performs a context switch-out as a result of a secure gateway.
+ *
+ * @internal
+ *
+ * This function is implemented as a wrapper, needed to make sure that the lr
+ * register doesn't get polluted and to provide context privacy during a context
+ * switch. The actual function is ::secure_gateway_context_switch_out. */
+void UVISOR_NAKED secure_gateway_out(uint32_t svc_sp)
+{
+    /* According to the ARM ABI, r0 will have the following value when this
+     * function is called:
+     *   r0 = svc_sp */
+    asm volatile(
+        "push {lr}\n"                                   /* Preserve the lr register. */
+        "bl   secure_gateway_context_switch_out\n"      /* secure_gateway_context_switch_out(svc_sp) */
```

Comment:
See [this comment](https://github.com/ARMmbed/uvisor/pull/218#discussion_r64188546).